File: leetcode/cpp/remove-linked-list-elements.cpp
#include <iostream>
using namespace std;

struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {} //构造函数初始化列表 http://www.cnblogs.com/bluetzar/articles/1223169.html
  };

/*方案一：用一个指针*/
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
		///*第一个while用于解决下面这个问题
		//Input:	[1], 1
		//Output:	{1}
		//Expected:	{}
        while (head != NULL && head->val == val)
        {
            head = head->next;
        }
		if (head == NULL) return NULL;
        ListNode *p;
		p = head;
		while(p->next!=NULL)
		{
			if(p->next->val==val)
			{
				p->next=p->next->next;
			}
			else
				p=p->next;
		}
		return head;
    }
};


/*方案二：用两个指针
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        while (head != NULL && head->val == val)
        {
            head = head->next;
        }
		if (head == NULL) return NULL;
        ListNode* prevPoint = head; // empty header
        ListNode* curPoint  = head->next; // the first valid node
        while(curPoint != NULL) {
            if(curPoint->val == val) { 
                prevPoint->next = curPoint->next; // unlink the node you remove
            }
            else {
                prevPoint = curPoint; 
                }
            curPoint = curPoint->next; // go to next value
        }
        return head;
    }
};*/

void print(ListNode *head)
{
    while (head != NULL)
    {
        cout<<head->val;
        head = head->next;
    }
    cout<<endl;
}

int main()
{
    Solution s;
    ListNode *node1 = new ListNode(1);
    ListNode *node2 = new ListNode(2);
    ListNode *node3 = new ListNode(1);
    ListNode *node4 = new ListNode(2);
    node1->next = node2;
    node2->next = node3;
    node3->next = node4;
    print(node1);
    node1 = s.removeElements(node1, 2);
    print(node1);
	system("pause");
}

File: leetcode/python/remove-linked-list-elements.py
import os

class Solution:
    # @param {ListNode} head
    # @param {integer} val
    # @return {ListNode}
    def removeElements(self, head, val):
        while (head != None) and (head.val == val):
            head = head.next
        if head == None:
            return None
        pre = head
        cur = head.next
        while cur != None:
            if cur.val == val:
                pre.next = pre.next.next
            else:
                pre = pre.next
            cur = cur.next

        return head

def printList(head):
    # @param {ListNode} head
    while head != None:
        print(head.val, end = ' ')
        head = head.next
    print() # print '\n'

if __name__ == '__main__':
    node1 = ListNode(1)
    node2 = ListNode(2)
    node3 = ListNode(1)
    node4 = ListNode(2)
    node1.next = node2
    node2.next = node3
    node3.next = node4

    printList(node1) # print the origin list
    Solver = Solution();
    node1 = Solver.removeElements(node1, 1)
    printList(node1) # print the result list

        